db/view/view.cc

```diff
@@ -3576,5 +3582,453 @@ sstring build_status_to_sstring(build_status status) {
+            }
+        }
+    }
+}
+
+future<> view_building_worker::update_building_state() {
+    auto guard = co_await _group0_client.start_operation(_as, service::raft_timeout{});
+    auto new_tasks = co_await _state.update(_db, _vb_state_machine.building_state);
+    co_await _state.finish_completed_tasks();
+    co_await start_new_tasks(std::move(new_tasks));
+}
+
+future<> view_building_worker::start_new_tasks(std::unordered_map<shard_id, std::vector<service::view_building::view_building_task>> new_tasks) {
+    for (auto& [shard, shard_tasks]: new_tasks) {
+        if (!is_shard_free(shard)) {
+            vbw_logger.warn("Tried to start new view building tasks on shard {} but the shard is busy", shard);
+            throw std::runtime_error(fmt::format("Shard {} is already busy", shard));
+        }
+
+        auto tasks = shard_tasks | std::views::transform([] (const service::view_building::view_building_task& t) {
+            return std::make_pair(t.id, t);
+        }) | std::ranges::to<std::unordered_map>();
+        auto batch = seastar::make_shared<view_building_worker::batch>(*this, tasks, shard_tasks.front().base_id, shard_tasks.front().replica);
+        for (auto& [id, _]: tasks) {
+            _state.tasks_map.insert({id, batch});
+        }
+        vbw_logger.debug("Started tasks {} for base table {} on shard {}", tasks | std::views::keys, shard_tasks.front().base_id, shard_tasks.front().replica.shard);
+        
+        co_await coroutine::maybe_yield();
+    }
+}
+
+bool view_building_worker::is_shard_free(shard_id shard) {
+    return !std::ranges::any_of(_state.tasks_map, [&shard] (auto& task_entry) {
+        return task_entry.second->replica.shard == shard;
+    });
+}
+
+void view_building_worker::init_messaging_service() {
+    ser::view_rpc_verbs::register_work_on_view_building_tasks(&_messaging, [this] (std::vector<utils::UUID> ids) -> future<std::vector<service::view_building::view_task_result>> {
+        return container().invoke_on(0, [ids = std::move(ids)] (view_building_worker& vbw) -> future<std::vector<service::view_building::view_task_result>> {
+            return vbw.get_tasks_results(std::move(ids));
+        });
+    });
+}
+
+future<> view_building_worker::uninit_messaging_service() {
+    return ser::view_rpc_verbs::unregister(&_messaging);
+}
+
+future<std::vector<service::view_building::view_task_result>> view_building_worker::get_tasks_results(std::vector<utils::UUID> ids) {
+    vbw_logger.debug("Got request for results of tasks: {}", ids);
+
+    // We assume that all tasks in `ids` are executed in the same batch.
+    // Group0_guard works as semaphore here and keeps view building state locked.
+    auto wait_until_batch_is_finished = [this, &ids] () -> future<service::group0_guard> {
+        auto id = ids.front();
+        
+        while (!_as.abort_requested()) {
+            auto guard = co_await _group0_client.start_operation(_as);
```

Comment:
This will perform a read barrier on every attempt. I think this is unnecessary after the first attempt. I suggest the following:

- Perform a single read barrier at the beginning of get_tasks_results, so that the call of this function sees the same state as the view build coordinator or newer.
- Instead of "starting the operation" every time you enter the loop, just acquire the read apply mutex (`auto guard = _group0_client.hold_read_apply_mutex()`).

The read apply mutex will make sure that the state of group0 won't change under your feet - in fact, if you take a look at the group0 guard struct you will see that it keeps this lock.